File: linked-list/mid_difficulty/143-reorder-list/19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(head == nullptr){
            return head;
        }

        if(head->next == nullptr && n == 1){
            return nullptr;
        }

        ListNode* temp = head;
        int len = 0;
        while(temp != nullptr){
            temp = temp->next;
            len++;
        }

        int nodeIndex = len - n;

        if(nodeIndex == 0){
            ListNode* toDelete = head;
            head = head->next;
            delete toDelete;
            return head;
        }

        ListNode* crr = head;
        ListNode* prev = head;
        int count = 0;
        while(crr != nullptr && count <= nodeIndex){
            if(count == nodeIndex){
                ListNode* toDelete = crr;
                prev->next = crr->next;
                crr = crr->next;
                delete toDelete;
            }else{
                prev = crr;
                crr = crr->next;
            }
            count++;
        }
        return head;
    }
};

File: Linked-list/19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null){
            return null;
        }

        int size = 0;
        ListNode crr = head;
        while(crr != null){
            crr = crr.next;
            size++;
        }

        crr = head;
        int indexToRemove = size - n;
        int iter = 1;

        if (indexToRemove == 0) {
            return head.next;
        }

        while(crr != null){
            if(crr.next != null && iter == indexToRemove){
                crr.next = crr.next.next;
                break;
            }
            crr = crr.next;
            iter++;
        }

        return head;
    }
}